### DISCUSSION TITLE
新手任务集合～

### DISCUSSION THREAD

**Discussion by chuntaojun** [polarismesh/polaris#355] [Show and tell]
## 如何发现新手任务呢？

可以在[**polarismesh**](https://github.com/polarismesh)这个组织下相关项目的issue列表中，查看打上了**good first issue**标签的

![企业微信截图_a2b2586f-aefa-4815-99f6-e2a172ca39b6](https://user-images.githubusercontent.com/23445172/165080854-d4ed09c5-2a2f-4e68-9ceb-07ecae216797.png)

## 如何认领任务

在对应的任务中回复 assigne me 即可

**Comment by chuntaojun** [comment]
# 服务发现实例推空保护

- [in review] polaris-java：https://github.com/polarismesh/polaris-java/issues/118
- polaris-go：https://github.com/polarismesh/polaris-go/issues/49
- polaris-cpp：https://github.com/polarismesh/polaris-cpp/issues/29

注意点

- 该特性需要做成开关形式

**Reply by gongzhijian** [reply]
assigne me

**Reply by chuntaojun** [reply]
> assigne me

polaris-java already have pepole assign，you can try polaris-go

**Reply by gongzhijian** [reply]
ok,polaris-java task aleady assigne me,thank you

**Reply by gongzhijian** [reply]
please code review @chuntaojun 涛哥

**Comment by chuntaojun** [comment]


**Reply by shichaoyuan** [reply]
assigne me

**Reply by michaelwangwangwang** [reply]
本质诉求是删除自动注册没有实例的服务列表，需要实时删除。也不需要定时去删除所有没有服务实例的服务列表。
能不能给自动创建的服务一个标志位，服务下线肯定会触发北极星server一些逻辑，顺便把这个下线服务的服务列表从控制台中删除，如果是没有自动注册服务的标志位，就不删除

**Reply by andrewshan** [reply]
是的，建议自动删除功能只针对自动创建的服务，如果用户手动创建的服务，则不需要走自动删除。
可以通过内部标签的方式来区分服务是自动创建还是手动创建，自动创建的服务默认打入internal-auto-created:true的标签，清理的时候，只针对带了改标签的服务。

**Reply by chuntaojun** [reply]
> 

这个在另一个优化PR中已经带入这个标签自动打上了

**Reply by shichaoyuan** [reply]
看到独立出了 polaris-cleanup 项目，考虑到后续可能会有很多自动清理的逻辑，这样是会好一些。

**Reply by shichaoyuan** [reply]
一个实现方案，大家给个意见
1. polaris-server 提供一个运维接口 GetEmptyService(filter)，用于查询空服务列表，filter 中可以传入 internal-auto-created:true 等条件；
2. cleanup 定时任务查询空服务列表，然后根据时长等条件判断是否需要调 DeleteServices 接口进行删除，是否删除的策略可以在 cleanup 进行灵活调整

**Reply by chuntaojun** [reply]
- 对于第一点，其实就直接用查询服务的接口就好了，然后多添加一个查询条件，即添加instanceEmpty=true 即可

**Reply by shichaoyuan** [reply]
嗯 复用console的接口也可以

**Comment by chuntaojun** [comment]


**Reply by chuntaojun** [reply]
is doing

**Comment by chuntaojun** [comment]
# 对于北极星server的metrics指标通过grafana进行展示

- polaris-server：https://github.com/polarismesh/polaris/issues/271

建议实现思路

- 了解北极星server当前暴露的指标
- 在grafana中将相关的指标进行图标展示
  - 对于图标需要有相关的注释信息，便于用户理解相关指标的含义

**Comment by chuntaojun** [comment]


**Reply by Albertchamberlain** [reply]
assigne me

**Comment by chuntaojun** [comment]


**Reply by shichaoyuan** [reply]
assigne me  自产自销吧

**Comment by chuntaojun** [comment]


**Reply by shichaoyuan** [reply]
assigne me
k8s新手尝试一下～

**Comment by andrewshan** [comment]
# nginx-gateway：支持kubernetes nginx能力，提供集成了北极星能力的nginx ingress镜像

nginx-gateway: https://github.com/polarismesh/nginx-gateway/issues/9

![image](https://user-images.githubusercontent.com/45474304/170811767-5bbe5993-9e3f-4832-bb45-236bc5a3a332.png)

目标：
通过polaris-controller接管k8s的服务，通过提供nginx ingress镜像，实现基于polaris服务发现的跨集群路由能力。

建议实现思路：
- 包装nginx-gateway，并打包成镜像，支持在k8s环境中拉起并进行API的分发

**Comment by chuntaojun** [comment]


**Reply by chuntaojun** [reply]
is doing

**Comment by chuntaojun** [comment]
# polaris-server 支持store层对接 pgsql 

polaris-server：https://github.com/polarismesh/polaris/issues/145

目标

- 通过新增一个store插件，在该插件内支持pgsql数据库的对接

建议实现思路

- 了解 store 层的接口定义
- 了解现有 sqldb store 插件的实现
- 仿照实现 pgsql 的 store 插件实现
  - 建议看下能不能走 orm 的形式，在这个新增的插件中，直接支持 pgsql、sqlserver 等数据库

**Comment by chuntaojun** [comment]
# 北极星内部事件中心插件机制优化

polaris-server：https://github.com/polarismesh/polaris/issues/342

目标

- 一个统一的事件中心
- 每种事件有自己单独的publisher，每种事件的产生、发布、消费互不影响
- 将原本的服务事件插件与该事件中心组合，建立一个Listener，该Listener的逻辑代替原本discoverevent的逻辑
- 后续基于此事件中心，实现配置相关事件的暴露、持久化日志等操作

建议实现思路

- 该组件放在 common package 中
- 符合 go 的编程理念
- 事件发布尽可能高效，不需要单独去创建EventProducer等
- 建议走方法注册的模式去注册EventConsumer
